ethpm_types/source.py

```diff
@@ -42,19 +43,43 @@ class Compiler(BaseModel):
     def __eq__(self, other) -> bool:
         if (
             not hasattr(other, "name")
-            and not hasattr(other, "version")
-            and not hasattr(other, "settings")
+            or not hasattr(other, "version")
+            or not hasattr(other, "settings")
         ):
             return NotImplemented
 
-        return (
-            self.name == other.name
-            and self.version == other.version
-            and self.settings == other.settings
-        )
+        return self.__hash__() == other.__hash__()
+
+    def _get_settings_str(self) -> str:
+        return self._stringify_settings(self.settings or {})
+
+    @classmethod
+    def _stringify_settings(cls, settings: Dict) -> str:
+        # NOTE: Exclude outputSelection as it may contain contract type names.
+        banned_fields = ("outputSelection",)
+        settings_filtered = {k: v for k, v in settings.items() if k not in banned_fields}
+
+        # For hashing and ID.
+        # Recursively sort dictionaries based on values
+        def sort_value(value: Any) -> Any:
+            if isinstance(value, dict):
+                return sort_dict(value)
+            elif isinstance(value, list):
+                return [sort_value(item) for item in value]
+            else:
+                return value
+
+        def sort_dict(_dict: Dict) -> Dict:
+            return {
+                k: sort_value(v)
+                for k, v in sorted(_dict.items(), key=lambda item: sort_value(item[1]))
+            }
```

Comment:
thought: double sort_value() calls could be costly depending on what we're talking about here.  Though building a list of tuples first might also suck.  :shrug: 

thought: Wonder if this is useful elsewhere enough to get moved to a util.  Deterministic object hashing has come up for me before in the past.

Reply:
i was thinking about util too.

another though i had for performance was just whitelisting values to include so we can just pass in a few keys that are common changes, like `evmVersion` and `optimizer`

Reply:
OK - so this comment made me think so I tried to address it anyway!

* Made the util for hashing strings.
* Got rid of double call to sort_value, wasnt needed, but still needed sorting keys of first dict.
* But bigger perf, made eq only check optimizer and evmVersion. In the future, can revisit a way to make more customizable but this at least solves the problem and isnt as costly.